pyscf/lib/np_helper/imatcopy.c

```diff
@@ -0,0 +1,380 @@
+{
+  const double complex alpha = *alphaptr;
+  const int nclean = n - n % TILESIZE;
+  const int ntiles = nclean / TILESIZE;
+
+#pragma omp parallel
+  {
+
+/*
+ *    The following loop nest is equivalent to:
+ *    for(int iouter = 1; iouter < ntiles; iouter++)
+ *      for(int jouter = 0; jouter < iouter; jouter++)
+ *
+ *    See 10.1109/IPDPS.2017.34. 
+ */
+    int first_iteration = 1;
+    int iouter, jouter;
+#pragma omp for nowait
+    for(long long ijouter = 0; ijouter < (ntiles*(ntiles-1))/2; ijouter++) {
+      if(first_iteration) {
+        iouter = uncollapse_loop_index(ijouter);
+        jouter = ijouter - iouter * (iouter - 1) / 2;
+        first_iteration = 0;
+      } else {
+        jouter++;
+        if(jouter == iouter) {
+          iouter++;
+          jouter = 0;
+        }
+      }
+      ztranspose_scale_tile_offdiag(A, iouter * TILESIZE, jouter * TILESIZE, lda, alpha);
+    }
+
+    if(alpha != 1.0) {
+#pragma omp for schedule(static) nowait
+      for(int ii = 0; ii < nclean; ii+=TILESIZE) {
+        ztranspose_scale_tile_diag(A, ii, ii, lda, alpha);
+      }
+    } else {
+#pragma omp for schedule(static) nowait
+      for(int ii = 0; ii < nclean; ii+=TILESIZE) {
+        ztranspose_tile_diag(A, ii, ii, lda);
+      }
+    }
+
+#pragma omp for schedule(static) nowait
+    for(int j = 0; j < nclean; j++) {
+      for(int i = nclean; i < n; i++) {
+        const double complex tmp = ELEM_AT(A, i, j, lda);
+        ELEM_AT(A, i, j, lda) = alpha * ELEM_AT(A, j, i, lda);
+        ELEM_AT(A, j, i, lda) = alpha * tmp;
+      }
+    }
+
+  } // end parallel region
+
+  for(int j = nclean; j < n; j++) {
+    for(int i = nclean; i < j; i++) {
+        const double complex tmp = ELEM_AT(A, i, j, lda);
+        ELEM_AT(A, i, j, lda) = alpha * ELEM_AT(A, j, i, lda);
+        ELEM_AT(A, j, i, lda) = alpha * tmp;
+    }
+  }
+
+  if(alpha != 1.0) {
+    for(int i = nclean; i < n; i++) {
+      ELEM_AT(A, i, i, lda) *= alpha;
+    }
+  }
+}
+
+
+
+/*
+ * Batched versions for 3D tensors
+ */
+
+void NPomp_dtensor_itranspose_scale021(const long long matstride, int nmat, int n, const double alpha,
+                                      double *A, int lda)
+{
+    if (omp_get_num_threads() >= nmat) {
+#pragma omp parallel
+{
+    int nlevels = omp_get_max_active_levels();
+    omp_set_max_active_levels(1);
+#pragma omp for schedule(static)
+    for (int imat = 0; imat < nmat; imat++) {
+      NPomp_d_itranspose_scale(n, alpha, A + imat * matstride, lda);
+    }
+    omp_set_max_active_levels(nlevels);
+}
+  } else {
+    for (int imat = 0; imat < nmat; imat++) {
+      NPomp_d_itranspose_scale(n, alpha, A + imat * matstride, lda);
+    }
+  }
+}
+
+void NPomp_ztensor_itranspose_scale021(const long long matstride, int nmat, int n, const double complex *alpha,
+                                      double complex *A, int lda)
+{
+    if (omp_get_num_threads() >= nmat) {
```

Comment:
Please keep the code simple. The basic OpenMP parallelization is sufficient here. Not necessary to introduce the nested omp directives.

Reply:
OK!